larky-api/src/main/java/com/verygood/security/larkyapi/VersionedLarkyEngineImpl.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    this.version = inputVersion;
+    URL larkyJarPath = larkyJarByVersion.get(version);
+    URLClassLoader childLoader = new URLClassLoader(
+            new URL[] {larkyJarPath},
+            VersionedLarkyEngineImpl.class.getClassLoader()
+    );
+
+    // equivalent to: import com.verygood.security.larky.jsr223.LarkyCompiledScript;
+    this.engineClass = Class.forName (
+            "com.verygood.security.larky.jsr223.LarkyScriptEngine",
+            true, childLoader);
+
+    // equivalent to: import com.verygood.security.larky.jsr223.LarkyScriptEngine;
+    this.scriptClass = Class.forName (
+            "com.verygood.security.larky.jsr223.LarkyCompiledScript",
+            true, childLoader);
+
+    // equivalent to: import com.verygood.security.larky.parser.ParsedStarFile;
+    this.parseClass = Class.forName (
+            "com.verygood.security.larky.parser.ParsedStarFile",
+            true, childLoader);
+
+    // create engine object
+    this.engineInstanceObj = (ScriptEngine) engineClass.newInstance();
+  }
+
+  @Override
+  public Object executeScript(String script,String outputVar)
+          throws NoSuchMethodException,IllegalAccessException,InvocationTargetException, ScriptException, NullPointerException {
+
+    Method compile = engineClass.getMethod("compile",String.class);
+    CompiledScript compiledScript = (CompiledScript) compile.invoke(engineInstanceObj, script);
+
+    Object starFile = compiledScript.eval();
+
+    Method getGlblVar = parseClass.getMethod (
+            "getGlobalEnvironmentVariable",
+            String.class,Class.class);
+    Object result = getGlblVar.invoke (starFile, outputVar, Object.class);
+
+    return result;
+  }
+
+  @Override
+  public Object executeScript(String script,String outputVar, ScriptContext context)
+          throws NoSuchMethodException,IllegalAccessException,InvocationTargetException,ScriptException {
+    setContext(context);
+    return executeScript(script, outputVar);
+  }
+
+  @Override
+  public Object executeScript(Reader script, String outputVar, ScriptContext context)
+          throws NoSuchMethodException,IllegalAccessException,InvocationTargetException,ScriptException  {
+    setContext(context);
+    return executeScript(script, outputVar);
+  }
+
+  @Override
+  public Object executeScript(Reader script, String outputVar)
+          throws NoSuchMethodException,IllegalAccessException,InvocationTargetException,ScriptException  {
+
+    Method compile = engineClass.getMethod("compile",Reader.class);
+    CompiledScript compiledScript = (CompiledScript) compile.invoke(engineInstanceObj, script);
+
+    Object starFile = compiledScript.eval();
+
+    Method getGlblVar = parseClass.getMethod(
+            "getGlobalEnvironmentVariable",
+            String.class,Class.class);
+    Object result = getGlblVar.invoke(starFile, outputVar, Object.class);
+
+    return result;
+  }
+
+  public static Set<String> getSupportedVersions() {
```

Comment:
`detectVersions` is being called on init and it can be called puclicly through `getSupportedVersions`.
Theoretically we can init with one set  of version and then on call for `getSupportedVersions` get different set of supported larky libs with rewritten map.
`getSupportedVersions` should be just informative w/o changing state of this object.

Reply:
I moved `detectVersions` into `static {...}` so this happens at class init. I think it should be better, let me know if you prefer something else